runtime/runtime-core/common/test/aws/smithy/kotlin/runtime/net/newnet/UrlTest.kt

```diff
@@ -0,0 +1,116 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package aws.smithy.kotlin.runtime.net.newnet
+
+import kotlin.test.Test
+import kotlin.test.fail
+import aws.smithy.kotlin.runtime.net.Url as OldUrl
+import aws.smithy.kotlin.runtime.net.newnet.Url as NewUrl
+
+class UrlTest {
+    private fun testEquivalence(vararg urls: String) {
+        val errors = urls.mapNotNull { url ->
+            val old = OldUrl.parse(url).toString()
+            val new = NewUrl.parseEncoded(url).toString()
+            if (old == new) null else "Old: <$old>, New: <$new>"
+        }
+        if (errors.isNotEmpty()) {
+            val message = errors.joinToString("\n", "Found the following errors:\n") { "  $it" }
+            fail(message)
+        }
+    }
+
+    @Test
+    fun testEquivalenceOfBasicUrls() {
+        testEquivalence(
+            "http://amazon.com",
+            "https://amazon.com",
+        )
+    }
+
+    @Test
+    fun testEquivalenceOfPorts() {
+        testEquivalence(
+            "https://amazon.com:8443",
+        )
+    }
+
+    @Test
+    fun testEquivalenceOfPaths() {
+        testEquivalence(
+            // "https://amazon.com/", // Old parser strips path `/`...is bug?
+            "https://amazon.com//",
+            "https://amazon.com/foo",
+            "https://amazon.com/foo/bar",
+            "https://amazon.com/foo/bar/",
+            "https://amazon.com/foo//bar/",
+            "https://amazon.com/foo/bar//",
+        )
+    }
+
+    @Test
+    fun testEquivalenceOfQueryStrings() {
+        testEquivalence(
+            // "https://amazon.com?", // Old parser strips '?' with empty query params...is bug?
+            "https://amazon.com?foo=bar",
+            // "https://amazon.com?foo=bar&baz=qux", // Old parser sorts query by key...is bug?
+            "https://amazon.com?foo=f%F0%9F%98%81o",
+            // "https://amazon.com?foo=f+o%20o", // Old parser encodes qparam ' ' to %20 instead of '+'...is bug?
+        )
+    }
```

Comment:
**Question**: No one remarked on these commented-out test cases. Our existing URL parser has some non-standard behavior that was surprising to me, such as stripping empty query strings and empty fragments. The new parser (correctly, I believe) retains them. This matches the RFC but is a different from our current behavior and may cause unexpected problems.

Do we have any thoughts about RFC adherence vis-à-vis current behavior?

Reply:
Stripping empty query strings and fragments 🤷 . I think a parser could set these as empty and if we go with my suggestion of a flag you could set the flag but in practice it's not really going to matter for us I don't think.

The `%20` vs `+` depends on whether the content type is `application/x-www-form-urlencoded`. 

https://en.wikipedia.org/wiki/Percent-encoding#The_application.2Fx-www-form-urlencoded_type

Our existing [implementation](https://github.com/awslabs/smithy-kotlin/blob/66e95fb45a9a508677eac67dcbd210820eb4dea7/runtime/runtime-core/common/src/aws/smithy/kotlin/runtime/net/QueryParameters.kt#L69) branches on this but from what I can tell we have generally not been using form-url encoding. Your decode function will have to handle this for query params but for encode I'd probably default to `%20` and maybe provide a flag or something to enable encoding as `+` like we do now.